py4DSTEM/process/classification/featurization.py

```diff
@@ -0,0 +1,637 @@
+        
+        Args
+        keys (list of str) list of feature names (same as in dict) to perform scaling on
+        """
+        for i in range(len(keys)):
+            rs = RobustScaler()
+            self.features[keys[i] + '_rs'] = rs.fit_transform(self.features[keys[i]])
+        return
+    
+    def shift_positive(self, keys):
+        """
+        Replaces a feature stored at feature.key with the positive shifted array.
+        
+        Args
+        keys (list of str) list of feature names (same as in dict) to perform shifting on
+        """
+        for i in range(len(keys)):
+            self.features[keys[i]] += np.abs(self.features[keys[i]].min())
+        return
+    
+    def PCA(self, keys, components):
+        """
+        Performs PCA on a set of keys
+        
+        Args:
+        keys (list of int, float, str): keys to performa PCA on
+        components (list): A list of ints for each key. This will be the output number of features
+        """
+        for i in range(len(keys)):
+            pca = PCA(n_components = components[keys[i]])
+            self.pca[keys[i]] = pca.fit_transform(self.features[keys[i]])
+        return
+
+    def ICA(self, keys, components):
+        """
+        Performs ICA on a set of keys
+        
+        Args:
+        keys (list of int, float, str): keys to performa ICA on
+        components (list): A list of ints for each key. This will be the output number of features
+        """
+        for i in range(len(keys)):
+            ica = FastICA(n_components = components[keys[i]])
+            self.ica[keys[i]] = ica.fit_transform(self.features[keys[i]])
+        return
+    
+    @ignore_warnings(category=ConvergenceWarning)
+    def NMF(self, keys, n_components, iters, random_state = None):
+        """
+        Performs traditional NMF on feature matrix.
+        
+        Args:
+        keys (list of str): List of strings associated with features
+        n_components (int): number of components (classes)
+        iters (int): number of iterations
+        random_state (int): random state
+        """   
+        if random_state == None:
+            rng = np.random.RandomState(seed = 42)
+        else:
+            seed = random_state
+        for i in range(len(keys)):
+            recon_error = np.inf
+            for n in range(len(iters)):
+                if random_state == None:
+                    seed = rng.randint(5000)
+                nmf_temp = NMF(n_components = n_components[keys[i]], random_state = seed)
+                nmf_ = nmf_temp.fit_transform(self.features[keys[i]])
+                if nmf_temp.reconstruction_err_ < recon_error:
+                    self.nmf[keys[i]] = nmf_
+                    self.nmf_comps[keys[i]] = nmf_temp.components_
+                    recon_error = nmf_temp.reconstruction_err_
+        return
+    
+    def NMF_iterative(self, keys, max_components, merge_thresh, iters, random_state = None, return_all = True):
```

Comment:
i'd suggest merging this function and the other `NMF` function into a single function.  If we set the `merge_thresh` default to some number greater than one, it should by default reduce to the normal `NMF` function